torch/csrc/generic/StorageSharing.cpp

```diff
@@ -160,7 +167,14 @@ static PyObject * THPStorage_(shareFd)(PyObject *_self, PyObject *noargs)
   } else {
     THWStoragePtr new_storage(
         THPStorage_(newFdStorage)(storage->nbytes() / sizeof(scalar_t)));
-    THWStorage_(copy)(new_storage, storage);
+
+    at::Storage new_storage_aten(
+      c10::intrusive_ptr<c10::StorageImpl>(
+        new_storage.get(),
+        c10::raw::DontIncreaseRefcount()));
```

Comment:
I think your refcounting issues stem from here, the `intrusive_ptr` needs to increase the refount. You can do this with `ATen/core/UnsafeFromTH.h`
```suggestion
    at::Storage new_storage_aten = at::unsafeStorageFromTH(new_storage.get(), /*retain=*/true);
```